### PR TITLE
:exploding_head: delete data specification

### PR DESCRIPTION
**Pull request by ccamel**
Specifies the `deleteData` operation that removes some RDF triples matching a where condition.

### CODE REVIEW

**Review by bdeneux — COMMENTED** [review]
I think maybe is missing a `DeleteResponse` that will return if possible the number of deleted triple or maybe the new store state (like is done on `StoreResponse`) ?

**Comment by ccamel** [comment]
> I think maybe is missing a `DeleteResponse` that will return if possible the number of deleted triple or maybe the new store state (like is done on `StoreResponse`) ?

🤔 ASAIK, for `Execute` messages in the contract, we use the `Response` type for which some [attributes](https://docs.rs/cosmwasm-std/latest/cosmwasm_std/struct.Response.html#method.add_attribute) are added. In that perspective, we can consider to add the number of deleted triples. But this is not explicitly described in the code (it could be in the description).

File: contracts/okp4-cognitarium/src/msg.rs [hunk 1]
```diff
@@ -28,6 +28,42 @@ pub enum ExecuteMsg {
...
+    /// Example:
+    /// ```json
+    /// {
+    ///   "prefixes": [
+    ///     { "prefix": "foaf", "namespace": "http://xmlns.com/foaf/0.1/" }
+    ///   ],
+    ///   "delete": [
+    ///     { "variable": "person" },
+    ///     { "variable": "property" },
+    ///     { "variable": "value" }
```
Comment (amimart) [inline contracts/okp4-cognitarium/src/msg.rs:45]: I don't understand the presence of the `property` and `value` variables in the example, is that an error?

File: contracts/okp4-cognitarium/src/msg.rs [hunk 2]
```diff
@@ -28,6 +28,42 @@ pub enum ExecuteMsg {
...
+    ///  ]
+    /// ```
+    ///
+    /// Only the smart contract owner (i.e. the address who instantiated it) is authorized to perform
+    /// this action.
+    DeleteData {
+        /// The prefixes used in the operation.
+        prefixes: Vec<Prefix>,
+        /// The items to delete.
+        delete: Vec<SelectItem>,
```
Comment (amimart) [inline contracts/okp4-cognitarium/src/msg.rs:62]: Just to check I understand it well, we delete the triples linked to the `SelectItems`? For example If a variable is present in the `WHERE` clause, but not selected, the underlying variable's triples won't be deleted, is that correct?

File: contracts/okp4-cognitarium/src/msg.rs (hunk 1 above)
Comment (ccamel) [reply contracts/okp4-cognitarium/src/msg.rs:45]: yes, well seen. It's a typo. Initially, I started by providing an example with 3 variables, but then reduced to a single variable. I will fix.

File: contracts/okp4-cognitarium/src/msg.rs (hunk 2 above)
Comment (ccamel) [reply contracts/okp4-cognitarium/src/msg.rs:62]: My bad, it's not correct. The item to delete should be `Vec<TriplePattern>`. This allows to specify the triples to delete by using a pattern with can refer to plain values or variables from the `where` clause.

**Review by amimart — CHANGES_REQUESTED** [review]
I just noticed some json issues in the example :)

File: contracts/okp4-cognitarium/src/msg.rs (hunk 2 above)
Comment (amimart) [reply contracts/okp4-cognitarium/src/msg.rs]: Ah this makes more sense, excellent! 😉

**Review by bdeneux — APPROVED** [review]
👍

**Pull request merged**